rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,211 @@
+- Start Date: 09-12-2022
+- Status: Proposed
+
+# Gateway refactor
+
+## Summary
+
+This proposes a full rewrite of the current gateway service. In addition, it also proposes
+changes to the way that access control to Jupyter servers is handled. Currently, each Jupyter
+server establishes a session with the user. But going forward each user will have a session
+with the gateway which will be used to control access into each Jupyter server. 
+The main motivation for changing the access control for Jupyter servers is the desire to run
+each Jupyter server in its own subdomain. The problem with unique subdomains is that each Jupyter
+server's subdomain would have to be registered in Keycloak so that the Oauth2-proxy in each 
+Jupyter server can function properly. But if the Oauth session is handled by the gateway we can
+have a single callback address in Keycloak as we currently do.
+
+## Motivation
+
+Reasons for rewriting the gateway:
+- the current code base is hard to maintain
+- one service could replace both traefik and the gateway
+- problems with managing sessions across different services/clients
+- using golang should result in more efficient use of resources and improved performance
+
+Reasons for centralized session authentication:
+- running sessions in subdomains which would allow for SSH access
+
+## Design Detail
+
+### Gateway
+
+The gateway has to have sessions with different clients. Some of these are:
+- User servers (i.e. the sidecar container needs be able to swap some kind of ID
+or access token for a Gitlab token)
+- Actual users logging in through the UI
+- Actual users logging in through the CLI
+
+Using JWT access tokens for this purpose is not feasible. This is because access
+tokens from Keycloak expire usually within an hour or several hours. This means
+that anyone who has to remain authenticated for a long time with the gateway
+needs to also get a refresh token and worry about using this to have valid access
+tokens. Accepting an expired access token is not secure.
```

Comment:
Maybe I've misunderstood, but what would be the problem with using the refresh token to get a new access token regularly?

Reply:
There can be a few issues with this:
- Some identity providers (keycloak by default does not do this) will make a refresh token invalid as soon as you have used it once. So by distributing refresh tokens to all services we get into a situation where we can have a race condition and have only a single service successfully refresh while all others are screwed.
- Distributing the refresh tokens in addition to the access tokens in many different places is less secure by storing them only in a single place and having only a single service access them.

Reply:
> How about having a single service responsible for refreshing the token then and storing the access token somewhere centralised like Vault instead?

That is exactly the idea.